File: livingrimoire start here/LivinGrimoire java/Auxiliary_Modules/WeightedResponder.java
package Auxiliary_Modules;

import java.util.ArrayList;
import java.util.List;
import java.security.SecureRandom;

public class WeightedResponder {
    private List<String> responses;
    private final int lim;

    // Constructor
    public WeightedResponder(int lim) {
        responses = new ArrayList<>();
        this.lim = lim;
    }

    // Method to get a response
    public String getAResponse() {
        int size = responses.size();
        if (size == 0) {
            return "";
        }

        int totalWeight = 0;
        int[] weights = new int[size];
        for (int i = 0; i < size; i++) {
            weights[i] = i + 1;
            totalWeight += weights[i];
        }
        int pick = new SecureRandom().nextInt(totalWeight);
        int cumulative = 0;
        for (int i = 0; i < size; i++) {
            cumulative += weights[i];
            if (pick < cumulative) {
                return responses.get(i);
            }
        }

        return responses.get(size - 1);
    }


    // Method to check if responses contain a string
    public boolean responsesContainsStr(String item) {
        return responses.contains(item);
    }

    // Method to check if a string contains any response
    public boolean strContainsResponse(String item) {
        for (String response : responses) {
            if (response.isEmpty()) {
                continue;
            }
            if (item.contains(response)) {
                return true;
            }
        }
        return false;
    }

    // Method to add a response
    public void addResponse(String s1) {
        if (responses.contains(s1)) {
            responses.remove(s1);
            responses.add(s1);
            return;
        }
        if (responses.size() > lim - 1) {
            responses.remove(0);
        }
        responses.add(s1);
    }
    public void addResponses(String... replies){
        for (String value : replies) {
            addResponse(value);
        }
    }
    public String getSavableStr() {
        return String.join("_", responses);
    }
    public String getLastItem() {
        if (responses.isEmpty()) {
            return "";
        }
        return responses.get(responses.size() - 1);
    }


    public WeightedResponder cloneObj() {
        WeightedResponder clonedResponder = new WeightedResponder(this.lim);
        clonedResponder.responses = new ArrayList<>(this.responses);
        return clonedResponder;
    }
}

File: livingrimoire start here/LivinGrimoire Kotlin/auxiliary_modules/WeightedResponder.kt
package auxiliary_modules

import java.security.SecureRandom

class WeightedResponder(private val lim: Int) {
    private var responses: MutableList<String> = ArrayList()

    val aResponse: String
        // Method to get a response
        get() {
            val size = responses.size
            if (size == 0) {
                return ""
            }
            var totalWeight = 0
            val weights = IntArray(size)
            for (i in 0 until size) {
                weights[i] = i + 1
                totalWeight += weights[i]
            }
            val pick = SecureRandom().nextInt(totalWeight)
            var cumulative = 0
            for (i in 0 until size) {
                cumulative += weights[i]
                if (pick < cumulative) {
                    return responses[i]
                }
            }
            return responses[size - 1]
        }

    // Method to check if responses contain a string
    fun responsesContainsStr(item: String): Boolean {
        return responses.contains(item)
    }

    // Method to check if a string contains any response
    fun strContainsResponse(item: String): Boolean {
        for (response in responses) {
            if (response.isEmpty()) {
                continue
            }
            if (item.contains(response)) {
                return true
            }
        }
        return false
    }

    // Method to add a response
    fun addResponse(s1: String) {
        if (responses.contains(s1)) {
            responses.remove(s1)
            responses.add(s1)
            return
        }
        if (responses.size > lim - 1) {
            responses.removeAt(0)
        }
        responses.add(s1)
    }

    fun addResponses(vararg replies: String) {
        for (value in replies) {
            addResponse(value)
        }
    }

    val savableStr: String
        get() = java.lang.String.join("_", responses)
    val lastItem: String
        get() = if (responses.isEmpty()) {
            ""
        } else responses[responses.size - 1]

    fun cloneObj(): WeightedResponder {
        val clonedResponder = WeightedResponder(lim)
        clonedResponder.responses = ArrayList(responses)
        return clonedResponder
    }
}